pkg/operator/ceph/object/realm/controller.go

```diff
@@ -0,0 +1,240 @@
+		if kerrors.IsNotFound(err) {
+			logger.Debug("CephObjectStoreRealm resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errors.Wrap(err, "failed to get CephObjectStoreRealm")
+	}
+
+	// The CR was just created, initializing status fields
+	if cephObjectStoreRealm.Status == nil {
+		cephObjectStoreRealm.Status = &cephv1.Status{}
+		cephObjectStoreRealm.Status.Phase = k8sutil.Created
+		err := opcontroller.UpdateStatus(r.client, cephObjectStoreRealm)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+		}
+	}
+
+	// Make sure a CephCluster is present otherwise do nothing
+	_, isReadyToReconcile, cephClusterExists, reconcileResponse := opcontroller.IsReadyToReconcile(r.client, r.context, request.NamespacedName, controllerName)
+	if !isReadyToReconcile {
+		// This handles the case where the Ceph Cluster is gone and we want to delete that CR
+		//
+		// Also, only remove the finalizer if the CephCluster is gone
+		// If not, we should wait for it to be ready
+		// This handles the case where the operator is not ready to accept Ceph command but the cluster exists
+		if !cephObjectStoreRealm.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Remove finalizer
+			err = opcontroller.RemoveFinalizer(r.client, cephObjectStoreRealm)
+			if err != nil {
+				return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+			}
+
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
+		return reconcileResponse, nil
+	}
+
+	// Set a finalizer so we can do cleanup before the object goes away
+	err = opcontroller.AddFinalizerIfNotPresent(r.client, cephObjectStoreRealm)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to add finalizer")
+	}
+
+	// DELETE: the CR was deleted
+	if !cephObjectStoreRealm.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("deleting realm CR %q", cephObjectStoreRealm.Name)
+
+		// Remove finalizer
+		err = opcontroller.RemoveFinalizer(r.client, cephObjectStoreRealm)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+		}
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	// validate the realm settings
+	err = ValidateRealmCR(cephObjectStoreRealm)
+	if err != nil {
+		cephObjectStoreRealm.Status.Phase = k8sutil.ReconcileFailedStatus
+		err := opcontroller.UpdateStatus(r.client, cephObjectStoreRealm)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+		}
+		return reconcile.Result{}, errors.Wrapf(err, "invalid name CR %q spec", cephObjectStoreRealm.Name)
+	}
+
+	// Start object reconciliation, updating status for this
+	cephObjectStoreRealm.Status.Phase = k8sutil.ReconcilingStatus
+	err = opcontroller.UpdateStatus(r.client, cephObjectStoreRealm)
```

Comment:
Wherever the status is updated, a recent pattern we added is to have an `updateStatus()` method that doesn't return any errors. It also has to get the latest CR before updating the status. See [this example](https://github.com/rook/rook/blob/8cdc3c103b77ad033dc1c355af0801562a5156dc/pkg/operator/ceph/cluster/controller.go#L502)